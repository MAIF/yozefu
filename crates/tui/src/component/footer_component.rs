//! The footer component displays contextual information: the current cluster, shortcuts and the last notifications
use crossterm::event::KeyEvent;

use ratatui::{
    layout::Rect,
    style::{Color, Style, Stylize},
    text::{Line, Span},
    Frame,
};
use tokio::sync::mpsc::UnboundedSender;

use crate::{
    action::{Action, Notification},
    error::TuiError,
};

use super::{Component, ComponentName, Shortcut, State};

#[derive(Default)]
pub struct FooterComponent {
    shortcuts: Vec<Shortcut>,
    main_component: ComponentName,
    state: Vec<ComponentName>,
    notification: Option<Notification>,
    action_tx: Option<UnboundedSender<Action>>,
    ticks: u64,
    show_shortcuts: bool,
}

impl FooterComponent {
    fn generate_shortcuts(&self, _state: &State) -> Vec<Span<'static>> {
        let mut spans = vec![];
        for shortcut in &self.shortcuts {
            spans.push(format!("[{}]", shortcut.key).bold());
            spans.push(format!(":{}   ", shortcut.description).into());
        }

        spans
    }

    pub fn show_shortcuts(&mut self, visible: bool) -> &Self {
        self.show_shortcuts = visible;
        self
    }
}

impl Component for FooterComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<(), TuiError> {
        self.action_tx = Some(tx);
        Ok(())
    }

    fn id(&self) -> ComponentName {
        ComponentName::Footer
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::Shortcuts(s, show) => {
                self.shortcuts = s;
                if show {
                    match self.main_component {
                        ComponentName::TopicsAndRecords => self
                            .shortcuts
                            .push(Shortcut::new("CTRL + O", "Hide topics")),
                        ComponentName::Records => self
                            .shortcuts
                            .push(Shortcut::new("CTRL + O", "Show topics")),
                        _ => (),
                    };
                }
                self.shortcuts.push(Shortcut::new("CTRL + H", "Help"));
                self.shortcuts.push(Shortcut::new("TAB", "Next panel"));
                self.shortcuts.push(Shortcut::new("ESC", "Quit"));
            }
            Action::ViewStack((main_component, views)) => {
                self.main_component = main_component;
                self.state = views;
            }
            Action::Notification(notification) => {
                self.ticks = 0;
                self.notification = Some(notification)
            }
            Action::ResetNotification() => {
                self.notification = None;
                self.ticks = 0;
            }
            Action::Tick => {
                self.ticks += 1;
                if self.ticks > 20 {
                    self.notification = None;
                }
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let mut view_stack = self.state.clone();
        view_stack.dedup();
        view_stack.push(state.focused.clone());
        let mut help = vec![];
        help.push(
            format!(" {} ", state.cluster)
                .black()
                .bold()
                .bg(state.theme.white),
        );
        help.push(" ".into());
        for v in view_stack.iter().enumerate() {
            let colors = match v.0 == view_stack.len() - 1 {
                true => (state.theme.bg_active, state.theme.fg_active),
                false => (state.theme.bg_disabled, state.theme.fg_disabled),
            };
            if v.0 > 0 {
                help.push("—".fg(colors.0));
            }
            let prefix = match v.0 {
                0 if self.main_component == ComponentName::TopicsAndRecords => "◧ ",
                0 if self.main_component == ComponentName::Records => "□ ",
                _ => "",
            };

            help.push(
                format!(" {}{:<8}", prefix, v.1.label())
                    .bg(colors.0)
                    .fg(colors.1)
                    .bold(),
            );
        }

        help.push(Span::from("  "));
        if self.show_shortcuts {
            help.extend(self.generate_shortcuts(state));
        }

        let line = Line::from(help);
        f.render_widget(line, rect);

        if let Some(n) = &self.notification {
            let notification =
                Span::styled(n.message.to_string(), Style::default().italic().not_bold());
            let r = Rect::new(
                rect.width
                    .saturating_sub(u16::try_from(n.message.len().checked_sub(3).unwrap_or(3))?),
                rect.y,
                n.message.len() as u16,
                1,
            );
            let notification = match n.level {
                log::Level::Error => notification.fg(Color::LightRed).underlined(),
                log::Level::Warn => notification.fg(Color::Yellow),
                _ => notification,
            };
            f.render_widget(notification, r);
        }
        Ok(())
    }
}
