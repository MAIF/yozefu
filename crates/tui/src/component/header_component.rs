//! The footer component displays contextual information: the current cluster, shortcuts and the last notifications
use crossterm::event::KeyEvent;

use ratatui::{
    Frame,
    layout::Rect,
    style::{Style, Stylize},
    text::{Line, Span},
};
use tokio::sync::mpsc::UnboundedSender;

use crate::{
    action::{Action, Notification},
    error::TuiError,
};

use super::{Component, ComponentName, State};

#[derive(Default)]
pub struct HeaderComponent {
    main_component: ComponentName,
    state: Vec<ComponentName>,
    notification: Option<Notification>,
    action_tx: Option<UnboundedSender<Action>>,
    ticks: u64,
}

impl Component for HeaderComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) {
        self.action_tx = Some(tx);
    }

    fn id(&self) -> ComponentName {
        ComponentName::Header
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::ViewStack((main_component, views)) => {
                self.main_component = main_component;
                self.state = views;
            }
            Action::Notification(notification) => {
                self.ticks = 0;
                self.notification = Some(notification)
            }
            Action::Tick => {
                self.ticks += 1;
                if self.ticks > 20 {
                    self.notification = None;
                }
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let mut view_stack = self.state.clone();
        view_stack.dedup();
        view_stack.push(state.focused.clone());
        let mut help = vec![];
        help.push(
            format!(" {} ", state.cluster)
                .black()
                .bold()
                .bg(state.theme.white),
        );
        help.push(" ".into());
        for v in view_stack.iter().enumerate() {
            let colors = match v.0 == view_stack.len() - 1 {
                true => (state.theme.bg_active, state.theme.fg_active),
                false => (state.theme.bg_disabled, state.theme.fg_disabled),
            };
            if v.0 > 0 {
                #[cfg(not(debug_assertions))]
                help.push("—".fg(colors.0));
                #[cfg(debug_assertions)]
                help.push("".to_string().bg(colors.0).fg(state.theme.bg));
            }
            let prefix = match v.0 {
                0 if self.main_component == ComponentName::TopicsAndRecords => "◧ ",
                0 if self.main_component == ComponentName::Records => "□ ",
                _ => "",
            };

            help.push(
                format!(" {}{:<8}", prefix, v.1.label())
                    .bg(colors.0)
                    .fg(colors.1)
                    .bold(),
            );
            #[cfg(debug_assertions)]
            help.push("".fg(colors.0));
        }

        help.push(Span::from("  "));

        let line = Line::from(help);
        f.render_widget(line, rect);

        if let Some(n) = &self.notification {
            let notification =
                Span::styled(n.message.to_string(), Style::default().italic().not_bold());
            let r = Rect::new(
                rect.width
                    .saturating_sub(u16::try_from(n.message.len().checked_sub(1).unwrap_or(1))?),
                rect.y,
                n.message.len() as u16,
                1,
            );
            let notification = match n.level {
                tracing::Level::ERROR => notification.fg(state.theme.red).underlined(),
                tracing::Level::WARN => notification.fg(state.theme.yellow),
                _ => notification,
            };
            f.render_widget(notification, r);
        }
        Ok(())
    }
}

#[cfg(test)]
use crate::assert_draw;

#[test]
fn test_draw() {
    let mut component = HeaderComponent::default();
    assert_draw!(component, 50, 3)
}
