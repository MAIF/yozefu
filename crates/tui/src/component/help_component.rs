//! Component showing the help

use crossterm::event::{KeyCode, KeyEvent};
use ratatui::{
    Frame,
    layout::Rect,
    style::Stylize,
    text::{Line, Span},
    widgets::{Block, BorderType, Borders, Clear, Padding, Paragraph, Wrap},
};

use crate::{Action, error::TuiError};

use super::{
    Component, ComponentName, Shortcut, State, issue_component::IssueComponent,
    scroll_state::ScrollState,
};

const HELP_HEIGHT: usize = 42;
const TEN_MINUTES_FRAME: usize = 30 * 60 * 10;
const REPOSITORY_URL: &str = concat!(
    "      https://github.com/MAIF/yozefu/tree/v",
    env!("CARGO_PKG_VERSION")
);

#[derive(Default)]
pub(crate) struct HelpComponent {
    scroll: ScrollState,
    rendered: usize,
}

impl Component for HelpComponent {
    fn id(&self) -> ComponentName {
        ComponentName::Help
    }

    fn shortcuts(&self) -> Vec<Shortcut> {
        vec![]
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        self.rendered = 0;
        match key.code {
            KeyCode::Char('k') | KeyCode::Down => {
                self.scroll.scroll_to_next_line();
            }
            KeyCode::Char('j') | KeyCode::Up => {
                self.scroll.scroll_to_previous_line();
            }
            KeyCode::Char('[') => {
                self.scroll.scroll_to_top();
            }
            KeyCode::Char(']') => {
                self.scroll.scroll_to_bottom();
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        f.render_widget(Clear, rect);

        let block = Block::default()
            .borders(Borders::ALL)
            .padding(Padding::horizontal(2))
            .border_type(BorderType::Rounded)
            .title(" Help ");

        let block = self.make_block_focused_with_state(state, block);

        let text = vec![
            Line::from(Span::raw("")),
            Line::from(Span::raw("")),
            Line::from("                                                           Key      Description").bold(),
            Line::from("                                                             /      Focus search input"),
            Line::from("                                                           ESC      Close the window/app"),
            Line::from("                                                           TAB      Focus next window"),
            Line::from("                                                   SHIFT + TAB      Focus previous window"),
            Line::from(""),

            Line::from("                                                      Variable      Type                        Alias       Description").bold(),
            Line::from(vec![Span::raw("                                                         topic      "), Span::from("String").fg(state.theme.green), Span::from("                          t").fg(state.theme.blue), Span::from("        Kafka topic")]),
            Line::from(vec![Span::raw("                                                        offset      "), Span::from("Number").fg(state.theme.green), Span::from("                          o").fg(state.theme.blue), Span::from("       Offset of the record")]),
            Line::from(vec![Span::raw("                                                           key      "), Span::from(""), Span::from("                                k").fg(state.theme.blue), Span::from("       Key of the record")]),
            Line::from(vec![Span::raw("                                                         value      "), Span::from(""), Span::from("                                v").fg(state.theme.blue), Span::from("       Value of the record")]),
            Line::from(vec![Span::raw("                                                     partition      "), Span::from("Number").fg(state.theme.green), Span::from("                          p").fg(state.theme.blue), Span::from("       Partition of the record")]),
            Line::from(vec![Span::raw("                                                     timestamp      "), Span::from("String").fg(state.theme.green), Span::from("                         ts").fg(state.theme.blue), Span::from("       Timestamp of the record (RFC 3339) â†’ 2025-06-01T12:00:00.000+02:00")]),
            Line::from(vec![Span::raw("                                                          size      "), Span::from("Number").fg(state.theme.green), Span::from("                         si").fg(state.theme.blue), Span::from("       Size of the record")]),
            Line::from(vec![Span::raw("                                                       headers      "), Span::from("Map<String, String>").fg(state.theme.green), Span::from("             h").fg(state.theme.blue), Span::from("       Headers of the record")]),
            Line::from(Span::raw("")),

            Line::from(vec![Span::from("                                                      Operator"), Span::from("      Type").fg(state.theme.green), Span::from("                                    Description").bold()]),
            Line::from(vec![Span::from("                                     == | != | > | >= | < | <="), Span::from("      Number | String").fg(state.theme.green), Span::from("                         Wayne's world, party time! Excellent!")]),
            Line::from(vec![Span::from("                                                 contains | ~="), Span::from("      String").fg(state.theme.green), Span::from("                                  Test if the variable contains the specified string")]),
            Line::from(vec![Span::from("                                                   starts with"), Span::from("      String").fg(state.theme.green), Span::from("                                  Test if the variable starts with the specified string")]),
            Line::from(""),


            Line::from(vec![Span::from("                                                        Clause      Syntax                                  Description").bold()]),
            Line::from(vec![Span::from("                                                         limit      limit <"), Span::from("number").fg(state.theme.yellow), Span::from(">                          Limit the number of kafka records to receive")]),
            Line::from(vec![Span::from("                                                          from      from <"), Span::from("begin").fg(state.theme.yellow), Span::from("|"), Span::from("end").fg(state.theme.yellow), Span::from("|"), Span::from("date").fg(state.theme.yellow), Span::from("|"), Span::from("offset").fg(state.theme.yellow), Span::from(">            Start consuming records from the beginning, the end or a date")]),
            Line::from(vec![Span::from("                                                      order by      order by <"), Span::from("var").fg(state.theme.yellow), Span::from("> <"), Span::from("asc").fg(state.theme.yellow), Span::from("|"), Span::from("desc").fg(state.theme.yellow), Span::from(">               Sort kafka records")]),
            Line::from(""),

            Line::from("                                                         Input      Description").bold(),
            Line::from(r#"                                    timestamp >= "1 hours ago"      All records published within the last hour"#),
            Line::from(r#"v contains "rust" and partition == 2 from beginning limit 1000      The first 1_000 kafka records from partition 2 containing 'rust' in the value"#),
            Line::from(r#"              (key == "ABC") || (key ~= "XYZ") from end - 5000      Among the latest 5_000 records, return the records where the key is "ABC" or the key contains "XYZ""#),
            Line::from(r#"                      value.hello == "world" order by key desc      Any kafka JSON record with a JSON property "hello" with the value "world", sorted by key in descending order"#),
            Line::from(""),
            Line::from(vec![
                Span::from("                                                         Theme").bold(),
                Span::from(format!(
                                        "      Theme is '{}'. run `yozf config get themes` to list available themes.",
                                        state.theme.name
                                    ))
            ]),
            Line::from(vec![
                Span::from("                                                 Configuration").bold(),
                Span::from(format!("      '{}'", state.configuration_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                          Logs").bold(),
                Span::from(format!("      '{}'", state.logs_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                       Filters").bold(),
                Span::from(format!("      '{}'", state.filters_dir.display()))
            ]),
            Line::from(vec![
                Span::from("                                                        Themes").bold(),
                Span::from(format!("      '{}'", state.themes_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                       Version").bold(),
                Span::from(REPOSITORY_URL)
            ]),
            Line::from(""),
        ];

        let paragraph = Paragraph::new(text)
            .wrap(Wrap { trim: false })
            .scroll((self.scroll.value(), 0));
        f.render_widget(paragraph.block(block), rect);

        if self.rendered > TEN_MINUTES_FRAME {
            let mut issue = IssueComponent::default();
            issue.draw(f, rect, state)?;
        }

        self.scroll.draw(f, rect, HELP_HEIGHT);
        self.rendered += 1;

        Ok(())
    }
}

#[cfg(test)]
use crate::assert_draw;

#[test]
fn test_draw() {
    let mut component = HelpComponent::default();
    assert_draw!(component, 300, 60)
}
