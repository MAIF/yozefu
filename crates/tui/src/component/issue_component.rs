use ratatui::{
    Frame,
    layout::{Constraint, Flex, Layout, Rect},
    style::{Style, Stylize},
    text::Line,
    widgets::{Block, BorderType, Borders, Clear, Padding, Paragraph},
};

use crate::error::TuiError;

use super::{Component, ComponentName, State};

#[derive(Default)]
pub(crate) struct IssueComponent {}

impl Component for IssueComponent {
    fn id(&self) -> ComponentName {
        ComponentName::Help
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let block = Block::default()
            .borders(Borders::ALL)
            .padding(Padding::symmetric(5, 0))
            .style(Style::default())
            .border_type(BorderType::Rounded);

        let block = self.make_block_focused(state, block);
        let paragraph = Paragraph::new(vec![
            Line::from(""),
            Line::from("ʕノ•ᴥ•ʔノ ︵ ┻━┻"),
            Line::from("Alt: Your mate flipping the desk").italic(),
            Line::from(""),
            Line::from("Are you struggling with the tool?"),
            Line::from("Leave us an issue so we can improve it:"),
            Line::from("https://github.com/MAIF/yozefu/issues").bold(),
            Line::from(""),
            Line::from("Press any key to close.").italic(),
        ]);

        let mut rect = rect;
        rect.width -= 2;
        rect.y += 1;

        let [area] = Layout::horizontal([Constraint::Length(54)])
            .flex(Flex::End)
            .areas(rect);
        let [area] = Layout::vertical([Constraint::Length(11)])
            .flex(Flex::Start)
            .areas(area);

        f.render_widget(Clear, area);
        f.render_widget(paragraph.block(block), area);

        Ok(())
    }
}
