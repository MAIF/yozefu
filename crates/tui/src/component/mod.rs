mod footer_component;
mod help_component;
mod issue_component;
mod progress_bar_component;
mod record_details_component;
mod records_component;
mod root_component;
mod schemas_component;
mod scroll_state;
mod search_component;
mod shortcut;
mod state;
mod topic_details_component;
mod topics_and_records_component;
mod topics_component;
pub mod ui;
mod vertical_scrollable_block;

use crossterm::event::{KeyEvent, MouseEvent};
use ratatui::{
    Frame,
    layout::Rect,
    style::{Style, Stylize},
    widgets::{Block, BorderType},
};
pub(crate) use root_component::RootComponent;
pub(crate) use shortcut::Shortcut;
use strum::Display;
use tokio::sync::mpsc::UnboundedSender;
pub use ui::Ui;

use std::sync::{Arc, LazyLock, Mutex};

pub use state::State;

use serde::Deserialize;

use crate::{Action, TuiError, records_buffer::RecordsBuffer, tui::Event};

pub(crate) type ConcurrentRecordsBuffer = LazyLock<Arc<Mutex<RecordsBuffer>>>;
static BUFFER: ConcurrentRecordsBuffer =
    LazyLock::new(|| Arc::new(Mutex::new(RecordsBuffer::new())));

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub(crate) enum FocusDirection {
    Top,
    Left,
    Right,
    Bottom,
}

#[derive(Debug, Clone, Display, Hash, PartialEq, Eq, Deserialize, PartialOrd, Ord)]
pub(crate) enum ComponentName {
    Records,
    Topics,
    Footer,
    RecordDetails,
    TopicsAndRecords,
    RecordsView,
    TopicDetails,
    Main,
    Search,
    Dialog,
    Help,
    Schemas,
}

impl ComponentName {
    pub fn label(&self) -> String {
        match &self {
            ComponentName::RecordDetails => "Record".to_string(),
            ComponentName::TopicDetails => "Topic".to_string(),
            _ => self.to_string(),
        }
    }
}

impl Default for ComponentName {
    fn default() -> Self {
        Self::Topics
    }
}

pub(crate) trait WithHeight: Component {
    fn content_height(&self) -> usize {
        0
    }
}

pub(crate) trait Component {
    fn register_action_handler(&mut self, _tx: UnboundedSender<Action>) {}

    fn id(&self) -> ComponentName;

    fn make_block_focused_with_state<'a>(&self, state: &State, block: Block<'a>) -> Block<'a> {
        match state.focused == self.id() {
            true => self.make_block_focused(state, block),
            false => block,
        }
    }

    fn make_block_focused<'a>(&self, state: &State, block: Block<'a>) -> Block<'a> {
        block
            .border_style(Style::default().fg(state.theme.focused_border))
            .border_type(BorderType::Thick)
            .title_style(Style::default().bold())
    }

    fn init(&mut self) -> Result<(), TuiError> {
        Ok(())
    }

    fn handle_events(&mut self, event: Option<Event>) -> Result<Option<Action>, TuiError> {
        let r = match event {
            Some(Event::Key(key_event)) => self.handle_key_events(key_event)?,
            Some(Event::Mouse(mouse_event)) => self.handle_mouse_events(mouse_event)?,
            _ => None,
        };
        Ok(r)
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn handle_mouse_events(&mut self, _mouse: MouseEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, _action: Action) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError>;

    fn shortcuts(&self) -> Vec<Shortcut> {
        vec![]
    }
}
