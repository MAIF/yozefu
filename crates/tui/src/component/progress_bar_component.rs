//! Progress bar at the top of the window when you consume kafka records.
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Style},
    widgets::Widget,
};

#[derive(Clone, Default)]
pub(crate) struct ProgressBarComponent {
    length: u64,
    progress: u64,
}

impl ProgressBarComponent {
    pub fn new(length: u64) -> Self {
        Self {
            length,
            progress: 0,
        }
    }

    pub fn set_progress(&mut self, inc: usize) {
        self.progress = inc as u64;
    }

    pub fn set_length(&mut self, length: usize) {
        self.length = length as u64;
    }
}

impl Widget for ProgressBarComponent {
    fn render(self, area: Rect, buf: &mut Buffer) {
        if self.progress == 0 || self.length == 0 {
            return;
        }
        let percent = 100 * self.progress / self.length;
        let dd = u64::from(area.right() - area.left()) * percent / 100;
        buf.set_string(
            area.left(),
            area.top(),
            (0..dd).map(|_| "â–”").collect::<String>(),
            Style::default().fg(Color::Green),
        );
    }
}

#[test]
fn test_draw() {
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    let mut component = ProgressBarComponent::default();
    let mut terminal = Terminal::new(TestBackend::new(50, 3)).unwrap();
    component.set_length(100);
    component.set_progress(30);
    terminal
        .draw(|frame| {
            frame.render_widget(component, frame.area());
        })
        .unwrap();
    assert_snapshot!(terminal.backend());
}
