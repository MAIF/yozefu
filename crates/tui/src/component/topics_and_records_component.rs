//! This component is a layout component that renders `[TopicsComponent]` and `[RecordsComponent]`.
use ratatui::{
    Frame,
    layout::{Constraint, Direction, Layout, Rect},
    widgets::Clear,
};
use std::sync::{Arc, Mutex};

use crate::{Action, component::topics_component::TopicsComponent, error::TuiError};

use super::{Component, ComponentName, State};

pub(crate) struct TopicsAndRecordsComponent {
    records: Arc<Mutex<dyn Component>>,
    topics: Arc<Mutex<TopicsComponent>>,
    longest_topic_size: u16,
}

impl TopicsAndRecordsComponent {
    pub fn new(topics: Arc<Mutex<TopicsComponent>>, records: Arc<Mutex<dyn Component>>) -> Self {
        let longest_topic_size = Self::longest_topics_length(topics.lock().unwrap().topics());
        Self {
            records,
            topics,
            longest_topic_size,
        }
    }

    fn longest_topics_length(topics: &[String]) -> u16 {
        topics
            .iter()
            .map(|s| s.len())
            .max()
            .unwrap_or(32)
            .try_into()
            .unwrap_or(32)
            .max(24)
    }

}

impl Component for TopicsAndRecordsComponent {
    fn id(&self) -> ComponentName {
        ComponentName::TopicsAndRecords
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        if let Action::Topics(new_topics) = action {
            self.longest_topic_size = Self::longest_topics_length(&new_topics);
        };
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        f.render_widget(Clear, rect);

        let chunks: std::rc::Rc<[Rect]> = Layout::default()
            .direction(Direction::Horizontal)
            .constraints([
                Constraint::Length(self.longest_topic_size + 10),
                Constraint::Percentage(100)
            ])
            .spacing(1)
            .split(rect);

        self.topics.lock().unwrap().draw(f, chunks[0], state)?;
        self.records.lock().unwrap().draw(f, chunks[1], state)?;
        Ok(())
    }
}

#[cfg(test)]
use crate::assert_draw;

#[cfg(test)]
#[test]
fn test_draw() {
    use lib::{DataType, KafkaRecord};
    use serde_json::json;

    use crate::component::{
        BUFFER, records_component::RecordsComponent, topics_component::TopicsComponent,
    };

    let topics_component = TopicsComponent::new(vec!["topic1".to_string()]);
    let records_component = RecordsComponent::new(&BUFFER);
    BUFFER.lock().unwrap().reset();
    BUFFER.lock().unwrap().push(KafkaRecord {
        topic: "movie-trailers".into(),
        timestamp: None,
        partition: 0,
        offset: 314,
        headers: Default::default(),
        key_schema: None,
        value_schema: None,
        size: 4348,
        key: DataType::String("7f12bd3b-4c96-4ba1-b010-8092234eec13".into()),
        key_as_string: "7f12bd3b-4c96-4ba1-b010-8092234eec13".into(),
        value: DataType::Json(json!(
            r#"{
            {
            "title" : "Swiss Army Man",
            "year": 20013
            }
            
            }"#
        )),
        value_as_string: Default::default(),
    });

    let mut component = TopicsAndRecordsComponent::new(
        Arc::new(Mutex::new(topics_component)),
        Arc::new(Mutex::new(records_component)),
    );
    assert_draw!(component, 120, 5)
}
