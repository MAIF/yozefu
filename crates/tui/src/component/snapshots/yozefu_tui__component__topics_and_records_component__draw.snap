---
source: crates/tui/src/component/topics_and_records_component.rs
expression: terminal.backend()
---
"┏ Topics [1] ━━┓ ╭ Records ────────────────────────────────────────────────────────────────────────────────────────────╮"
"┃ [x] topic1   ┃ │Timestamp                             Topic   Offset         Key  Value                              │"
"┃              ┃ │                                                                                                     │"
"┃              ┃ │1970-01-01T01:00:00.000+01:00  movi..ers[0]      314  7f12bd3b..                                     │"
"┗━━━━━━━━━━━━━━┛ ╰─────────────────────────────────────────────────────────────────────────────────────────────────────╯"
